src/Collector.cs

```diff
@@ -0,0 +1,195 @@
+      // If you want the worker loop, look at SendQueuedMessages
+      while (!_worker.CancellationPending)
+        {
+          try
+            {
+              socket = Connect();
+              Authenticate(socket);
+              SendQueuedMessages(socket);
+              CloseSocket(socket);
+              failures = 0;
+            }
+          catch (Exception e)
+            {
+              _log.Error("An exception occurred", e);
+              if (socket != null)
+                {
+                  try {
+                    CloseSocket(socket);
+                  } catch {}
+                  socket = null;
+                }
+              var delay = (int) Math.Min(MaxReconnectDelay, Math.Pow(failures++, Backoff));
+              _log.ErrorFormat("Disconnected. {0} failures in a row. Reconnect in {1} seconds.", failures, delay);
+              Thread.Sleep(delay*1000);
+            }
+        }
+    }
+
+    private void SendQueuedMessages(Socket socket)
+    {
+      while (!_worker.CancellationPending)
+        {
+          // only pop if the last _currentCommand did not send
+          if (_currentCommand == null)
+            _currentCommand = _messages.Take();
+
+          if (!IsSocketConnected(socket))
+            throw new Exception("Disconnected");
+
+          _log.DebugFormat("Sending: {0}", _currentCommand);
+          var data = System.Text.Encoding.ASCII.GetBytes(_currentCommand + "\n");
+
+          socket.Send(data);
+          _currentCommand = null;
+        }
+    }
+
+    private void Authenticate(Socket socket)
+    {
+      var helloString = $"hello version dotnet/{Agent.AgentVersion}\n";
+      var authenticateString = $"authenticate {_apiKey}\n";
+
+      var data = System.Text.Encoding.ASCII.GetBytes(helloString + authenticateString);
+      socket.Send(data);
+
+      //hello ok?
+      if(!ReceiveOk(socket)) throw new Exception("Instrumental Authentication Failed");
+
+      //authenticate ok?
+      if(!ReceiveOk(socket)) throw new Exception("Instrumental Authentication Failed");
+    }
+
+    private bool ReceiveOk(Socket socket)
+    {
+      byte[] buffer = new byte[3];
+      socket.Receive(buffer);
+      // I'm not including all of LINQ for this shit
+      // return InstrumentalOk.SequenceEqual(buffer);
+      return InstrumentalOk[0] == buffer[0] && InstrumentalOk[1] == buffer[1] && InstrumentalOk[2] == buffer[2];
+    }
+
+    private static Socket Connect()
+    {
+      var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+      _log.Info("Connecting to collector.");
+      socket.Connect(InstrumentalAddress, InstrumentalPort);
+      _log.Info("Connected to collector.");
+      return socket;
+    }
+
+    private static bool IsSocketConnected(Socket socket)
+    {
+      bool eitherHasDataOrIsDead = socket.Poll(1000, SelectMode.SelectRead);
+      bool hasNoData = (socket.Available == 0);
+
+      return !(eitherHasDataOrIsDead && hasNoData);
```

Comment:
I suspect it would not shock you to learn that I tried many, many versions of this.

I think your version is clearer than this version, doing !Poll() really opens up the landscape.